Fix init container update validation for pods

ivan4th opened the pull request:
Partial fix #26840

The remaining issues with `kubectl apply` on pods with init containers
are caused by temporary annotation-based representation and
will resolve themselves once init containers leave alpha state.
Also, this PR makes sure internal and external objects don't get mixed up by the
PATCH handler (see related issue #25106).

This PR is an alternative for #28557 which met criticism from @smarterclayton 
and @liggitt for working around the temporary issue with annotations.
#28557 is a full fix for #26840 and contains an e2e test that cannot pass

without the `VolumeMounts` workaround. As there appears to be no
good way to include an e2e test that's known to be failing in k8s source,
I've removed it from this PR.

Either this PR or #28557 should be applied, but not both.

ivan4th committed:
Fix update validation for pods w/init containers.

liggitt commented on pkg/apiserver/resthandler_test.go:
```diff
@@ -217,7 +220,7 @@ func (tc *patchTestCase) Run(t *testing.T) {
 
 		case api.StrategicMergePatchType:
-			patch, err = strategicpatch.CreateStrategicMergePatch(originalObjJS, changedJS, versionedObj)
+			patch, err = strategicpatch.CreateTwoWayMergePatch(originalObjJS, changedJS, tc.startingPod)
```
the point of this was to test StrategicMergePatch...

liggitt replied:
still use `versionedObj`

liggitt commented on pkg/apiserver/resthandler_test.go:
```diff
@@ -261,10 +265,6 @@ func (tc *patchTestCase) Run(t *testing.T) {
-		if err != nil {
-			t.Errorf("%s: unexpected error: %v", tc.name, err)
-			return
-		}
```
we want to check this bit

liggitt commented on pkg/apiserver/resthandler_test.go:
```diff
@@ -187,12 +196,6 @@ func (tc *patchTestCase) Run(t *testing.T) {
 	copier := runtime.ObjectCopier(api.Scheme)
 	resource := unversioned.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"}
 
-	versionedObj, err := api.Scheme.ConvertToVersion(&api.Pod{}, unversioned.GroupVersion{Version: "v1"})
```
I think we just want `versionedObj := &apiv1.Pod{}`

liggitt commented on pkg/apiserver/resthandler_test.go:
```diff
@@ -324,6 +324,54 @@ func TestPatchResourceWithVersionConflict(t *testing.T) {
+	tc.changedPod.UID = uid
+	tc.changedPod.ResourceVersion = "1"
+	tc.changedPod.APIVersion = "v1"
+	tc.changedPod.Spec.DeprecatedServiceAccount = "acct"
```
I'm not seeing any difference between startingPod and changedPod... won't that make a no-op patch?

liggitt commented on pkg/apiserver/resthandler_test.go:
```diff
@@ -149,12 +150,12 @@ type patchTestCase struct {
 
 	// startingPod is used as the starting point for the first Update
-	startingPod *api.Pod
+	startingPod *v1.Pod
```
not sure what the point of switching all of these to v1.Pod is... the runtime.Encode was already ensuring we had v1 JSON as our patches. What the codec does (even when encoding a v1.Pod to JSON) is opaque enough that if you want to really test something like the deprecated service account field, I'd suggest setting up a separate test that actually uses JSON strings instead of relying on a codec step inside the test loop

liggitt commented on pkg/api/validation/validation.go:
```diff
@@ -1960,28 +1960,44 @@ func ValidatePodSecurityContext(securityContext *api.PodSecurityContext, spec *a
 	return allErrs
 }
 
+func validateContainerUpdates(oldContainers, newContainers []api.Container, fldPath *field.Path) (allErrs field.ErrorList, stop bool) {
```
keep the new,old ordering consistent with ValidatePodUpdate

ivan4th commented:
@liggitt updated the PR.
1. In `strategicpatch.go`, we have
```
// Deprecated: Use the synonym CreateTwoWayMergePatch, instead.
func CreateStrategicMergePatch(original, modified []byte, dataStruct interface{}) ([]byte, error) {
	return CreateTwoWayMergePatch(original, modified, dataStruct)
}
```
That's why I've replaced `CreateStrategicMergePatch` with `CreateTwoWayMergePatch`.

2. As of dataStruct argument, it's used in the following manner in `CreateTwoWayMergePatch`:
```
	t, err := getTagStructType(dataStruct)
```
and `getTagStructType()` just does `reflect.TypeOf(dataStruct)`, so because `tc.startingPod` is `v1.Pod` after my changes making another `v1.Pod` to pass as `dataStruct` (`versionedObj`) doesn't make any sense.

3. `startingPod` & `changedPod`: the naming is a bit confusing but if you look at the other tests in the same file you'll see that `updatedPod` corresponds to the new 'client-side' version, while `changedPod` refers to the server-side pod that was probably changed, too. Still, I removed the mixup test because indeed the 'mixup' testing should be done in `apiserver_test.go` (but I'm not sure it's worth the effort).

4. I've modified resthandler test to use v1.Pod objects because `patchResource()` usually accepts v1.Pod objects during normal API request handling. I think this way it's less confusing for people trying to understand k8s encoding/decoding/conversion/defaulting code which can be quite confusing as #25106 shows. But this part is not critical, may revert it back to api.Pod if you think it would be better to keep it this way.

5. Fixed missing error check & argument order inconsistency.

liggitt commented:
> making another v1.Pod to pass as dataStruct (versionedObj) doesn't make any sense

passing test data into that function is misleading... passing an empty `&v1.Pod{}` in is clearer

liggitt commented:
I'd rather leave the patch tests using api.Pod, as they are today... switching to v1.Pod doesn't make it much clearer what data is being operated on, since those still have to be passed through a codec to get JSON.

If you want to test things like patching init containers in JSON (which I think could be worthwhile), I would set up a different patch test that takes actual JSON strings to represent the data submitted by the user and the data read from etcd. That makes it crystal clear what the data being operated on is.

liggitt commented:
the init container validation change (and associated tests), and removing api.Pod from grouplessGroupVersion make sense to me. If you want to test init container annotation patches, let's do that via JSON tests without changing the existing patch tests.

ivan4th commented:
Removed resthandler_test changes other than simplification of `versionedObj`. I think I'll make a separate PR for JSON-based patch testing.

liggitt commented on pkg/api/validation/validation.go:
```diff
@@ -1960,28 +1960,44 @@ func ValidatePodSecurityContext(securityContext *api.PodSecurityContext, spec *a
-	}
 
 	// validate updateable fields:
 	// 1.  containers[*].image
```
update comment to include `initContainers[*].image`. @smarterclayton, I assume initContainers image fields are supposed to be updateable as well?

liggitt commented:
nit on comment, LGTM otherwise

smarterclayton commented:
Yes, initContainer image fields may be updated as regular containers.

On Tue, Jul 19, 2016 at 5:26 PM, Jordan Liggitt <notifications@github.com>
wrote:

> nit on comment, LGTM otherwise
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29034#issuecomment-233771252>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p7kkNzVuq7-eDKkYlOwi2ho8rlA9ks5qXUD3gaJpZM4JN1ez>
> .
>

ivan4th commented:
Updated the comment.
There's also [an error message](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L2024) mentioning `containers[*]` a few lines below it but as currently the user must use annotations for init containers it's too early too update the user-visible message.

liggitt commented:
LGTM
